### PR TITLE
Implement joint trajectory player

### PR DESCRIPTION
**Pull request by isorrentino**
The application reads a trajectory from a file and executes it either on a simulated robot or on a real robot. It also saves the measured joint positions. It also allows you to specify in a configuration file which robot joints to use.

### CODE REVIEW

File: utilities/joint-trajectory-player/include/BipedalLocomotion/JointTrajectoryPlayer/Module.h [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+{
+    double m_dT; /**< RFModule period. */
+    std::string m_robot; /**< Robot name. */
+
+    Eigen::VectorXd m_currentJointPos;
+
+    std::shared_ptr<yarp::dev::PolyDriver> m_robotDevice;
+
+    RobotInterface::YarpRobotControl m_robotControl;
+    RobotInterface::YarpSensorBridge m_sensorBridge;
```
Comment (GiulioRomualdi) [inline utilities/joint-trajectory-player/include/BipedalLocomotion/JointTrajectoryPlayer/Module.h:40]: Could you please add the documentation here?

File: utilities/joint-trajectory-player/src/Module.cpp [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+#include <BipedalLocomotion/RobotInterface/YarpRobotControl.h>
+#include <BipedalLocomotion/RobotInterface/YarpSensorBridge.h>
+
+#include <Eigen/Dense>
+
+#include <BipedalLocomotion/JointTrajectoryPlayer/Module.h>
+
+#include <yarp/dev/IEncoders.h>
+
+using Vector1d = Eigen::Matrix<double, 1, 1>;
```
Comment (GiulioRomualdi) [inline utilities/joint-trajectory-player/src/Module.cpp:22]: I think you do not need this

File: utilities/joint-trajectory-player/src/Module.cpp [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+                  << std::endl;
+        return false;
+    }
+
+    // check the number of controlled joints
+    m_numOfJoints = 0;
+    yarp::dev::IEncoders* axis;
+    m_robotDevice->view(axis);
+    if (axis != nullptr)
+        axis->getAxes(&m_numOfJoints);
```
Comment (GiulioRomualdi) [inline utilities/joint-trajectory-player/src/Module.cpp:54]: Is there another way to get this information in the `RobotInterface`?

File: utilities/joint-trajectory-player/src/Module.cpp [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+    {
+        std::cerr << "[Module::initializeSensorBridge] Unable to set the driver list" << std::endl;
+        return false;
+    }
+
+    return true;
+}
+
+std::pair<bool, std::deque<Eigen::VectorXd>>
+Module::readStateFromFile(const std::string& filename, const std::size_t num_fields)
```
Comment (GiulioRomualdi) [inline utilities/joint-trajectory-player/src/Module.cpp:107]: `num_fields` -> `numFields`

File: utilities/joint-trajectory-player/src/Module.cpp [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+
+bool Module::updateModule()
+{
+    bool isMotionDone;
+    bool isMimeExpired;
+    std::vector<std::pair<std::string, double>> jointlist;
+
+    m_robotControl.checkMotionDone(isMotionDone, isMimeExpired, jointlist);
+
+    if (isMotionDone)
```
Comment (GiulioRomualdi) [inline utilities/joint-trajectory-player/src/Module.cpp:224]: This function was not meant to be used in this way. What I did in the past is to create a state machine and use `checkMotionDone` as a trigger to change the state. We can discuss this offline

File: utilities/joint-trajectory-player/src/Module.cpp [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+    // set the file name
+    std::time_t t = std::time(nullptr);
+    std::tm tm = *std::localtime(&t);
+
+    std::ofstream stream; /**< std stream. */
+    std::stringstream fileName;
+
+    fileName << "Dataset_Measured_" << m_robotControl.getJointList().front() << "_"
+             << std::put_time(&tm, "%Y_%m_%d_%H_%M_%S") << ".txt";
+    stream.open(fileName.str().c_str());
```
Comment (GiulioRomualdi) [inline utilities/joint-trajectory-player/src/Module.cpp:275]: If I'm not mistaken the main purpose of this application is to collect some data to analyze them in matlab. In this case you may also use `matio-cpp` developed by @S-Dafarra. You can find a usage example in  https://github.com/dic-iit/bipedal-locomotion-framework/pull/62

**Review by GiulioRomualdi — CHANGES_REQUESTED** [review]
I would also add a readme where you could explain how to run the application

**Comment by isorrentino** [comment]
I modified the code based on @GiulioRomualdi suggestions. The PR is again ready for review.

**Review by GiulioRomualdi — APPROVED** [review]
Thanks @isorrentino

File: utilities/joint-trajectory-player/src/Module.cpp (hunk 6 above)
Comment (prashanthr05) [reply utilities/joint-trajectory-player/src/Module.cpp]: Standalone short Example usage: for reading and writing.

https://github.com/dic-iit/element_floating-base-estimation/blob/feature/contacts/code/external-cpp/FKContacts/src/LeggedOdomTest.cpp

**Pull request merged**